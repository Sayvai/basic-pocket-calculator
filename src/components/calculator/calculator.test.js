import React from 'react';
import { fireEvent, render, screen } from '@testing-library/react';
import Calculator from './calculator';

describe('Calculator component', () => {
  let config;

  beforeEach(() => {
    config = {
      brandName: 'Custom Brand',
      assetUrl: 'https://www.someurl.com/logo.svg',
      brandingColorHex: '#1795d4',
      backgroundColorHex: '#1795d4',
    };
  });

  describe('renderings', () => {
    it('should render the child components <Branding>, <Screen>, and <Buttons>, with the correct props', () => {
      // when
      const { container } = render(<Calculator config={config} />);

      // then
      expect(screen.getByTestId('component-branding')).toBeDefined();
      expect(screen.getByTestId('component-screen')).toBeDefined();
      expect(screen.getByTestId('component-buttons')).toBeDefined();

      expect(container).toMatchSnapshot();
    });
  });

  describe('button interactions', () => {
    it('should initially render screen with no content', () => {
      // when
      render(<Calculator config={config} />);

      // then
      const screenOutput = screen.getByTestId('component-screen-input');
      expect(screenOutput.value).toBe('');
    });

    it('should render button key value entry onto the screen', () => {
      // given
      render(<Calculator config={config} />);

      // when
      const buttonEight = screen.getByTestId('Eight');
      const screenOutput = screen.getByTestId('component-screen-input');

      fireEvent.click(buttonEight);

      // then
      expect(screenOutput.value).toBe('8');
    });

    it('should clear the screen after a button key value entry', () => {
      // given
      render(<Calculator config={config} />);

      // when
      const buttonFive = screen.getByTestId('Five');
      const buttonClear = screen.getByTestId('Clear');
      const screenOutput = screen.getByTestId('component-screen-input');

      fireEvent.click(buttonFive);

      // then
      expect(screenOutput.value).toBe('5');

      // when
      fireEvent.click(buttonClear);

      // then
      expect(screenOutput.value).toBe('');
    });

    it('should render the correct calculated result for 1+1', () => {
      // given
      render(<Calculator config={config} />);

      // when
      const buttonOne = screen.getByTestId('One');
      const buttonAdd = screen.getByTestId('Add');
      const buttonEqual = screen.getByTestId('Equal');
      const screenOutput = screen.getByTestId('component-screen-input');

      fireEvent.click(buttonOne);
      fireEvent.click(buttonAdd);
      fireEvent.click(buttonOne);

      // then
      expect(screenOutput.value).toBe('1+1');

      // when
      fireEvent.click(buttonEqual);

      // then
      expect(screenOutput.value).toBe('2');
    });
  });
});