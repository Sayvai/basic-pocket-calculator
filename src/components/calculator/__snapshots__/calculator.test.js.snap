// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calculator component renderings should render the child components <Branding>, <Screen>, and <Buttons>, with the correct props 1`] = `
<div>
  <div
    class="calculator"
    data-testid="component-caculator"
    style="background-color: rgb(23, 149, 212);"
  >
    <div
      class="branding"
      data-testid="component-branding"
      style="background-color: rgb(23, 149, 212);"
    >
      <img
        alt="Custom Brand logo"
        src="https://www.someurl.com/logo.svg"
      />
      <h3>
        Pocket Calculator
      </h3>
    </div>
    <div
      class="screen"
      data-testid="component-screen"
    >
      <input
        class="output"
        data-testid="component-screen-input"
        readonly=""
        value=""
      />
    </div>
    <div
      class="buttons"
      data-testid="component-buttons"
    >
      <button
        aria-label="Clear"
        class="button-clear"
        data-button-value="c"
        data-testid="Clear"
      >
        C
      </button>
      <button
        aria-label="Divide"
        data-button-operator="true"
        data-button-value="/"
        data-testid="Divide"
      >
        ÷
      </button>
      <button
        aria-label="Seven"
        data-button-value="7"
        data-testid="Seven"
      >
        7
      </button>
      <button
        aria-label="Eight"
        data-button-value="8"
        data-testid="Eight"
      >
        8
      </button>
      <button
        aria-label="Nine"
        data-button-value="9"
        data-testid="Nine"
      >
        9
      </button>
      <button
        aria-label="Multiply"
        data-button-operator="true"
        data-button-value="*"
        data-testid="Multiply"
      >
        ×
      </button>
      <button
        aria-label="Four"
        data-button-value="4"
        data-testid="Four"
      >
        4
      </button>
      <button
        aria-label="Five"
        data-button-value="5"
        data-testid="Five"
      >
        5
      </button>
      <button
        aria-label="Six"
        data-button-value="6"
        data-testid="Six"
      >
        6
      </button>
      <button
        aria-label="Subtract"
        data-button-operator="true"
        data-button-value="-"
        data-testid="Subtract"
      >
        −
      </button>
      <button
        aria-label="One"
        data-button-value="1"
        data-testid="One"
      >
        1
      </button>
      <button
        aria-label="Two"
        data-button-value="2"
        data-testid="Two"
      >
        2
      </button>
      <button
        aria-label="Three"
        data-button-value="3"
        data-testid="Three"
      >
        3
      </button>
      <button
        aria-label="Add"
        data-button-operator="true"
        data-button-value="+"
        data-testid="Add"
      >
        +
      </button>
      <button
        aria-label="Zero"
        class="button-zero"
        data-button-value="0"
        data-testid="Zero"
      >
        0
      </button>
      <button
        aria-label="Equal"
        class="button-equal"
        data-button-operator="true"
        data-button-value="="
        data-testid="Equal"
      >
        =
      </button>
    </div>
  </div>
</div>
`;
